### ISSUE TITLE
Required reference with a nullable FK doesn't throw errors in sqlite and returns null for containing objects in queries

### ISSUE THREAD

**Issue by mikamikem** [dotnet/efcore#27411]
The SQLite provider doesn't throw any errors during model creation in the following scenario and generates null values for the object containing the required reference when querying (even though running the logged SQL produced by the query in the example on the database outside of EF Core shows proper values being joined), which helped hide where the real issue was for a while.  Curiously (as I found out when trying to distill the issue down), my specific use case seems to have hit a very precise gap in the error handling because moving the reference around to other owned models in the hierarchy causes proper error messages on model creation or when adding entities even though that seems like it should be unrelated.  I've documented those in comments in the code.

### The sample code:

```C#
public class BaseType
{
    public Guid IdGuid { get; set; }

    public FirstLayerOwned FirstLayerSingle;
}

public class FirstLayerOwned
{
    public List<SecondLayerOwned> SecondLayerCollection;
}

public class SecondLayerOwned
{
    public ThirdLayerOwned ThirdLayerSingle;
}

public class ThirdLayerOwned
{
    // Moving all of these fields to FirstLayerOwned throws an error on
    // model creation and moving it to SecondLayerOwned also throws an
    // error on model creation.  For some reason putting it here doesn't...
    private BaseType baseRef;

    public BaseType BaseRef
    {
        get
        {
            return (BaseType)(object)baseRef;
        }
        set
        {
            baseRef = (BaseType)(object)value;

            baseRefGuid = value?.IdGuid;
        }
    }

    public Nullable<Guid> baseRefGuid;

    public string Test { get; set; }
}

public class OwnedEntityContext : DbContext
{
    public DbSet<BaseType> BaseValues { get; set; }

    protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
        => optionsBuilder
            .UseSqlite(@"Filename=Test.sqlite");

    protected override void OnModelCreating(ModelBuilder modelBuilder)
    {
        modelBuilder.Entity<BaseType>().Property(baseType => baseType.IdGuid)
                                       .IsRequired();
        modelBuilder.Entity<BaseType>().HasKey(baseType => baseType.IdGuid);

        modelBuilder.Entity<BaseType>().OwnsOne(
            b => b.FirstLayerSingle,
            firstBuilder =>
            {
                firstBuilder.WithOwner();

                firstBuilder.OwnsMany(
                    first => first.SecondLayerCollection,
                    secondBuilder =>
                    {
                        secondBuilder.WithOwner();

                        // Removing the RowID related calls causes an error to be flagged during model building
                        secondBuilder.Property<int>("RowId");

                        secondBuilder.HasKey("RowId");

                        secondBuilder.OwnsOne(
                            second => second.ThirdLayerSingle,
                            thirdBuilder =>
                            {
                                thirdBuilder.WithOwner();

                                thirdBuilder.HasOne(fifth => fifth.BaseRef)
                                            .WithMany()
                                            // Everything works fine if you remove this
                                            .IsRequired()
                                            .HasForeignKey(baseType => baseType.baseRefGuid);

                                thirdBuilder.Property(fifth => fifth.Test)
                                            .UsePropertyAccessMode(PropertyAccessMode.Property);
                            });
                    });
            });
    }
}

private static void Main(string[] args)
{
    #region Init data
    using (var context = new OwnedEntityContext())
    {
        context.Database.EnsureDeleted();
        context.Database.EnsureCreated();

        context.Add(new BaseType
        {
            FirstLayerSingle = new FirstLayerOwned
            {
                SecondLayerCollection = new List<SecondLayerOwned>
                    {
                        new SecondLayerOwned
                        {
                            ThirdLayerSingle = new ThirdLayerOwned
                            {
                                BaseRef = null,

                                Test = "This is a test string."
                            }
                        }
                    }
            }
        });

        context.SaveChanges();
    }
    #endregion

    using (var context = new OwnedEntityContext())
    {
        var baseType = context.BaseValues
            .Include(baseType => baseType.FirstLayerSingle)
            .Include(baseType => baseType.FirstLayerSingle)
                .ThenInclude(first => first.SecondLayerCollection)
            .Include(baseType => baseType.FirstLayerSingle)
                .ThenInclude(first => first.SecondLayerCollection)
                    .ThenInclude(second => second.ThirdLayerSingle).First();

        // This line throws a NullRefException trying to access ThirdLayerSingle
        Console.WriteLine($"First test string is: {baseType.FirstLayerSingle.SecondLayerCollection[0].ThirdLayerSingle.Test}");
    }
}
```

Switching to the SqlServer provider results in an error being thrown on model creation and moving the BaseRef field around to different classes in the hierarchy causes the SQLite provider to also throw errors on model creation.  If BaseRef has a valid value, this code actually works as expected, which makes it even more concerning and also made tracking down the source of the issue difficult.

It would be great if this would get an error message in model creation like it does on the SqlServer provider so that it's relatively straightforward to track down.

I noticed two other issues were already logged related to cascade delete functionality changing in EF Core 6, but I think this is unrelated because 1) this was happening in 5.x as well and 2) the values are still present in the database if you use another tool to view it, they just aren't properly retrieved through EF Core.

### Provider and version information

EF Core version: 6.0.2
Database provider: Microsoft.EntityFrameworkCore.Sqlite - 6.0.2
Target framework: .NET 6.0
Operating system: Windows 10
IDE: Visual Studio 2022 17.0.1

**Comment by AndriySvyryd** [comment]
**TLDR**: Add this to `OnModelCreating`:
```C#
secondBuilder.Navigation(s => s.ThirdLayerSingle).IsRequired();
```

**Explanation**: There are two things happening at once here. By default owned entities are considered optional, so if they share the table with the owner their columns are nullable even if the corresponding properties are required.
When querying EF determines whether an optional owned entity exists by checking its required properties, if there are any that are `null` then the owned entity cannot exist and therefore not materialized.

The exception on SqlServer is unrelated and if fixed it behaves the same.
